File: 0509-fibonacci-number/0509-fibonacci-number.py
class Solution:
    def fib(self, n: int, memo=None) -> int:
        if memo is None:
            memo = {0: 0, 1: 1}
        
        if n not in memo:
            memo[n] = self.fib(n-1, memo) + self.fib(n-2, memo)
        
        return memo[n]

File: 0509-fibonacci-number/0509-fibonacci-number.rs
impl Solution {
    pub fn fib(n: i32) -> i32 {
        if n < 2 {
        return n;
        }
        let mut fib_vec = Vec::with_capacity(4 * (n as usize + 1));
        fib_vec.push(0);
        fib_vec.push(1);
        for i in 2..=n as usize {
            fib_vec.push(fib_vec[i - 1] + fib_vec[i - 2]);
        }
        fib_vec[n as usize]
    }
}